### PR TITLE
Support for rendering Soccer Scenes using Bedlam characters

### PR DESCRIPTION
**Pull request by hakanardo**
This has become a quite large change, which still needs a bit of cleaning up, but I'm creating this pull-request now to get the discussion going on how we want it merged. Squash it all to one big commit? Keep the messy history for reference? Or something in between?

### CODE REVIEW

File: Makefile
Comment (KatjaPalmkvist) [inline Makefile:1]: "Streamline installation of optional custom plugins" was cherry picked to pull req #3 and should probably be removed from this one?

Comment (hakanardo) [reply Makefile:1]: Right! I'll fix...

File: Makefile [hunk 1]
```diff
@@ -114,6 +120,10 @@ constraints.txt: $(wildcard requirements/*.txt)
 docker-build:
 	docker build -t blenderset .
 
+docker-push: docker-build
+	docker tag blenderset hakanardo/blenderset
+	docker push hakanardo/blenderset
```
Comment (glajsnake) [inline Makefile:125]: Move to a variable at the top instead of hardcoding here?

File: Makefile
Comment (glajsnake) [inline Makefile:1]: -mkdir -p $(@D)/$(BLENDER_VERSION)/python/include/python3.10

Should we use the nice PYTHON_VERSION variable instead?

File: blenderset/assets.py [hunk 2]
```diff
@@ -81,6 +85,29 @@ def get_all_objects_of_class(self, class_name):
                 objs.append(obj)
         return objs
 
+    def get_roi(self):
+        if self.override_roi is not None:
+            return self.override_roi
+        walkable_polys = self.get_all_proprty_values("blenderset.visible_walkable_roi")
+        return MultiPolygon([Polygon(p) for p in walkable_polys])
+
```
Comment (glajsnake) [inline blenderset/assets.py:93]:
```suggestion
# Returns a random point within the region of interest (roi) if present, else return a random point between +- 1 m from origin of the 3d model.
```

File: blenderset/assets.py [hunk 3]
```diff
@@ -81,6 +85,29 @@ def get_all_objects_of_class(self, class_name):
...
+    def random_position(self, roi):
+        if roi.bounds and all(np.isfinite(roi.bounds)):
+            min_x, min_y, max_x, max_y = roi.bounds
+            while True:
```
Comment (glajsnake) [inline blenderset/assets.py:97]: Possible infinite loop if the polygon is bad. Please add break condition or change to a for loop.

File: blenderset/background.py [hunk 4]
```diff
@@ -292,12 +292,15 @@ def __init__(
         self, context, blend_file, camera_name=None, apply_background_modifiers=True
     ):
         super().__init__(context)
-        self.blend_file = str(self.root / "background_models" / blend_file)
+        if isinstance(blend_file, str):
```
Comment (glajsnake) [inline blenderset/background.py:295]: Please add short documentation for intended use of blend_files, whether root should be included or not depends on if it is a string or list.

**Review by glajsnake — COMMENTED** [review]
Nice work!
Just a few comments on clarifications.

**Comment by hakanardo** [comment]
I believe all raised issues now to be resolved.

File: blenderset/background.py (hunk 4 above)
Comment (glajsnake) [reply blenderset/background.py:302]: Thank you for the documentation.
If I remember correctly, the isinstance(blend_file, str) is to be backwards compatible and should not be used?
From the lines below it looks like the blend_file list should actually contain the entire path including self.root / "background_models" or wherever the files are located? If so we would recommend updating "filenames" to "file paths" in the documentation above. 

(also some typos: "whoes" -> "whos" and "backgournd" -> "background")

As it is clear that blend_file should be a list from the description, perhaps we can rename the variable to blend_files? Or does that break the backwards compatibility for you?

**Review by glajsnake — COMMENTED** [review]
Thank you for the update! 
We're just wondering about a minor detail.

File: blenderset/background.py (hunk 4 above)
Comment (hakanardo) [reply blenderset/background.py]: Good idea! Should be fixed now.

**Review by glajsnake — APPROVED** [review]
Thank you, looks good!

**Pull request merged**